src/wallet/migrate.cpp

```diff
@@ -0,0 +1,771 @@
+        s >> uint32_flags;
+        if (other_endian) {
+            uint32_flags = internal_bswap_32(uint32_flags);
+        }
+        flags = static_cast<BTreeFlags>(uint32_flags);
+
+        s >> uid;
+        s >> unused2;
+        s >> minkey;
+        s >> re_len;
+        s >> re_pad;
+        s >> root;
+        s >> unused3;
+        s >> crypto_magic;
+        s >> trash;
+        s >> iv;
+        s >> chksum;
+
+        if (other_endian) {
+            lsn = internal_bswap_64(lsn);
+            page_num = internal_bswap_32(page_num);
+            magic = internal_bswap_32(magic);
+            version = internal_bswap_32(version);
+            pagesize = internal_bswap_32(pagesize);
+            free_list = internal_bswap_32(free_list);
+            last_page = internal_bswap_32(last_page);
+            partitions = internal_bswap_32(partitions);
+            key_count = internal_bswap_32(key_count);
+            record_count = internal_bswap_32(record_count);
+            unused2 = internal_bswap_32(unused2);
+            minkey = internal_bswap_32(minkey);
+            re_len = internal_bswap_32(re_len);
+            re_pad = internal_bswap_32(re_pad);
+            root = internal_bswap_32(root);
+            crypto_magic = internal_bswap_32(crypto_magic);
+        }
+
+        // Page number must match
+        if (page_num != expected_page_num) {
+            throw std::runtime_error("Meta page number mismatch");
+        }
+
+        // Check magic
+        if (magic != BTREE_MAGIC) {
+            throw std::runtime_error("Not a BDB file");
+        }
+
+        // Only version 9 is supported
+        if (version != 9) {
+            throw std::runtime_error("Unsupported BDB data file version number");
+        }
+
+        // Page size must be 512 <= pagesize <= 64k, and be a power of 2
+        if (pagesize < 512 || pagesize > 65536 ||  (pagesize & (pagesize - 1)) != 0) {
+            throw std::runtime_error("Bad page size");
+        }
+
+        // Page type must be the btree type
+        if (type != PageType::BTREE_META) {
+            throw std::runtime_error("Unexpected page type, should be 9 (BTree Metadata)");
+        }
+
+        // Only supported meta-flag is subdatabase
+        if (flags != BTreeFlags::SUBDB) {
+            throw std::runtime_error("Unexpected database flags, should only be 0x20 (subdatabases)");
+        }
+    }
+};
+
+/** General class for records in a BDB BTree database. Contains common fields. */
+class RecordHeader
+{
+public:
+    uint16_t len;  // Key/data item length
+    RecordType type;  // Page type and DELETE FLAG
+
+    static constexpr size_t SIZE = 3; // The record header is 3 bytes
+
+    bool other_endian;
+
+    RecordHeader(bool other_endian) : other_endian(other_endian) {}
+    RecordHeader() = delete;
+
+    RecordType GetRealType() const
+    {
+        return static_cast<RecordType>(static_cast<uint8_t>(type) & ~static_cast<uint8_t>(RecordType::DELETE));
```

Comment:
Instead of casting back and forth and having a `GetRealType()` it would, imo, be more straightforward to split the delete flag off from the type in `Unserialize` before casting to `RecordType`. So to have
```
RecordType type;
bool deleted;
```
Then these accessors could go.

Reply:
Done as suggested.